### PR TITLE
Expose the function to spawn invalidation on cdn

### PR DESCRIPTION
**Pull request by Spring3**
I think that this is the only way to do this and not end up refactoring the whole project.
Since everything is bound and is made private, I can not expose the invalidation function, because it is non standard and might be AWS specific (so there might be no reason in exposing it in the `index.js` file.

### CODE REVIEW

File: src/providers/s3.js [hunk 1]
```diff
@@ -65,11 +74,36 @@ class S3Provider extends BaseProvider {
...
 
+  invalidate(invalidationPaths = ['/*'], distribution = process.env.CLOUDFRONT_DISTRIBUTION_ID) {
+    return new Promise((resolve, reject) => {
+      assert(distribution, 'distribution has to be a non empty string');
+      assert(
+        (invalidationPaths && typeof invalidationPaths === 'string')
+        || Array.isArray(invalidationPaths),
+        'invalidationPaths must be a string or array of strings'
+      );
+      const targetPaths = Array.isArray(invalidationPaths) ? invalidationPaths : [invalidationPaths];
```
Comment (mrister) [inline src/providers/s3.js]: you already did a check for Array.isArray(invalidationPaths) reuse it

File: test/intergration/provider-s3.js [hunk 2]
```diff
@@ -34,7 +35,7 @@ describe('S3 Provider', () => {
       const stream = fs.createReadStream('./LICENSE');
       return manager
         .uploadFile(testLocation, testFileName, stream)
-        .then((result) => {
+        .then(({ result }) => {
```
Comment (mrister) [inline test/intergration/provider-s3.js]: hm this is a potential breaking change we would need to publish a major version

**Review by mrister — COMMENTED** [review]
Implementation seems ok but I'm not sure about the way it does it because it has a potential breaking change due to alteration of the output of the functions, perhaps another approach would be good, ti expose a file-manager-invalidation related class from the package and use it separately, with some examples in the redme.

**Review by jkernech — APPROVED** [review]

**Review by ofkindness — APPROVED** [review]

File: src/providers/s3.js (hunk 1 above)
Comment (Spring3) [reply src/providers/s3.js]: I am not sure I understand how I can reuse it. The previous check is inside the assertion to validate if it's either a string or an array. Then, I convert a string to array to make it easier to process later

**Comment by Spring3** [comment]
@mrister I requested your review again as you are the initial author of this module.
I loosened a bit the structure of the exposed object from the factory.create() and instead of exposing a strict number of functions, I just expose the whole provider. I think that this way it is both not having any breaking changes and allows some custom functions within the provider (except the default uploadFile, downloadFile, deleteFile). Why I did that is because `invalidate` might not be a function that each provider can provide (let's say S3 can, but dropbox for example can not etc). This way we won't suddenly bump into the "Not implemented" kind of error in runtime

**Review by ykhatal — APPROVED** [review]

File: src/index.js [hunk 3]
```diff
@@ -11,43 +11,7 @@ const defaultConfig = require('./providers/config').create();
...
-     * @method
-     * @name deleteFile
-     * @param args
-     * @return {Promise}
-     */
-    deleteFile(...args) {
-      return activeProvider.deleteFile(...args);
-    }
-  };
+  return activeProvider;
```
Comment (mrister) [inline src/index.js]: if everything is removed then there is no need for activeProvider variable it ca be returned at the beginning

File: src/providers/s3.js (hunk 1 above)
Comment (mrister) [reply src/providers/s3.js]: hm something like this then:
```js
const isArray = Array.isArray(invalidationPaths);
assert( typeof invalidationPaths === 'string' || isArray)
const targetPaths = isArray ? invlidationPaths : [invalidationPaths]
```

File: README.md [hunk 4]
```diff
@@ -43,6 +43,24 @@ const manager = fileManagement.create('S3', {
       .catch(console.error);
 ```
 
+#### Run cloudfront invalidation
+After `upload` function is executed, it exposes `invalidate` function along with the original upload result:
+```js
+manager.uploadFile(...args)
+.then(({ invalidate, result }) => {
```
Comment (mrister) [inline README.md]: this is still a breaking change for this version, we then need a new major version or not put invalidation here at all

**Review by mrister — COMMENTED** [review]
@Spring3 can we discuss why the approach here is to bundle invalidation with the uploadFile function and why it cannot be a separate module in this package for someone to use as it pleases and still keep the minor version?

File: README.md (hunk 4 above)
Comment (Spring3) [reply README.md]: I forgot to update readme. it does not work like this anymore

**Review by mrister — APPROVED** [review]
Ah ok invalidate is a separate top level function. Update readme with example and perfect.

**Pull request merged**